src/main/java/picard/sam/BamErrorMetric/ReadBaseStratification.java

```diff
@@ -0,0 +1,909 @@
+        NS_IN_READ(() -> nsInReadStratifier, "Stratifies bases according to the number of Ns in the read.");
+
+        private final String docString;
+
+        @Override
+        public String getHelpDoc() {
+            return docString + " Suffix is " + stratifier.get().getSuffix();
+        }
+
+        private final Supplier<RecordAndOffsetStratifier<?>> stratifier;
+
+        Stratifiers(final Supplier<RecordAndOffsetStratifier<?>> stratifier, final String docString) {
+            this.stratifier = stratifier;
+            this.docString = docString;
+        }
+
+        public RecordAndOffsetStratifier<?> makeStratifier() {
+            return stratifier.get();
+        }
+    }
+
+    /**
+     * An enum to hold the ordinality of a read
+     */
+    public enum ReadOrdinality {
+        FIRST,
+        SECOND;
+
+        public static ReadOrdinality of(final SAMRecord sam) {
+            if (!sam.getReadPairedFlag()) {
+                return null;
+            }
+            return sam.getFirstOfPairFlag() ? ReadOrdinality.FIRST : ReadOrdinality.SECOND;
+        }
+    }
+
+    /**
+     * An enum designed to hold a binned version of
+     * any probability-like number (between 0 and 1)
+     * in quintiles
+     */
+    public enum CycleBin {
+        // CycleBins are groupings of inputs by their binned quintile (20%)
+        // Each bin represents 20% of the distribution.
+        QUINTILE_1(0, 0.2),
+        QUINTILE_2(0.2, 0.4),
+        QUINTILE_3(0.4, 0.6),
+        QUINTILE_4(0.6, 0.8),
+        QUINTILE_5(0.8, 1.0);
+
+        final double lower;
+        final double upper;
+
+        CycleBin(double lower, double upper) {
+            this.lower = lower;
+            this.upper = upper;
+        }
+
+        static CycleBin valueOf(double value) {
+            return Stream.of(CycleBin.values())
+                    .filter(e -> value >= e.lower && value <= e.upper)
+                    .findFirst()
+                    .orElseThrow(() -> new IllegalArgumentException(String.format("Value for CycleBin must be between 0 and 1 (inclusive), found: %g", value)));
+        }
+    }
+
+    /**
+     * An enum for holding the direction for a read (positive strand or negative strand
+     */
+    public enum ReadDirection {
+        POSITIVE("+"),
+        NEGATIVE("-");
+
+        private final String outputString;
+
+        ReadDirection(final String output) {
+            outputString = output;
+        }
+
+        @Override
+        public String toString() {
+            return outputString;
+        }
+
+        public static ReadDirection of(final SAMRecord sam) {
+            return sam.getReadNegativeStrandFlag() ? ReadDirection.NEGATIVE : ReadDirection.POSITIVE;
+        }
+    }
+
+    /**
+     * An enum for holding a reads read-pair's Orientation (i.e. F1R2, F2R1, F1F2 or R1R2)
+     */
+    public enum PairOrientation {
+        F1R2,
+        F2R1,
+        F1F2,
+        R1R2;
```

Comment:
Why not use `PairOrientation` in htsjdk?

Reply:
Pair orientation in htsJdk is Innie vs. Outtie. vs. Tandem, which is not what I have here.

Reply:
See your doc comment, which _could_ be interpreted differently. Change `F1R2, F2R1, F1F2 or R1R2` to `F1R2, F2R1, F1F2, or R1R2`